src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+// only recomputes data when the hashes change, otherwise it computes
+// all non null scene data.
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraphEvaluatorFilter : public tbb::filter
+{
+	public:
+		SceneGraphEvaluatorFilter( const ScenePlug *scene, const Context *context, bool onlyChanged ) :
+			tbb::filter( tbb::filter::parallel ), m_scene( scene ), m_context( context ), m_onlyChanged( onlyChanged )
+		{
+		}
+
+		void *operator()( void *item )
+		{
+			SceneGraph *s = (SceneGraph*)item;
+			ScenePlug::ScenePath path;
+			s->path( path );
+
+			try
+			{
+				ContextPtr context = new Context( *m_context, Context::Borrowed );
+				context->set( ScenePlug::scenePathContextName, path );
+				Context::Scope scopedContext( context.get() );
+			
+				if( m_onlyChanged )
+				{
+					// we're re-traversing this location, so lets only recompute attributes where
+					// their hashes change:
+
+					IECore::MurmurHash attributesHash = m_scene->attributesPlug()->hash();
+					if( attributesHash != s->m_attributesHash )
+					{
+						s->m_attributesHash = attributesHash;
+						s->m_attributes = m_scene->attributesPlug()->getValue();
+					}
+				}
+				else
+				{
+					// Compute the initial attribute hash, and all the data that hasn't already
+					// been computed:
+
+					s->m_attributesHash = m_scene->attributesPlug()->hash();
+					if( !s->m_attributes )
+					{
+						s->m_attributes = m_scene->attributesPlug()->getValue();
+					}
+					if( !s->m_object )
+					{
+						s->m_object = m_scene->objectPlug()->getValue();
+					}
+					if( !s->m_transform )
+					{
+						s->m_transform = new IECore::M44fData( m_scene->transformPlug()->getValue() );
+					}
+				}
+			}
+			catch( const std::exception& e )
+			{
+				std::string name;
+				ScenePlug::pathToString( path, name );
+			
+				IECore::msg( IECore::Msg::Error, "InteractiveRender::update", name + ": " + e.what() );
+			}
+
+			return s;
+		}
+
+	private:
+
+		const ScenePlug *m_scene;
+		const Context *m_context;
+		const bool m_onlyChanged;
+};
+
+//////////////////////////////////////////////////////////////////////////
+// SceneGraphOutputFilter implementation
+//
+// This serial thread bound filter outputs scene data to a renderer on
+// the main thread, then discards that data to save memory. If the
+// editMode flag is set to true, the filter outputs edits, otherwise
+// it renders the data directly.
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraphOutputFilter : public tbb::thread_bound_filter
+{
+	public:
+	
+		SceneGraphOutputFilter( Renderer *renderer, bool editMode ) :
+			tbb::thread_bound_filter( tbb::filter::serial_in_order ), 
+			m_renderer( renderer ),
+			m_attrBlockCounter( 0 ),
+			m_editMode( editMode )
+		{
+		}
+		
+		~SceneGraphOutputFilter()
+		{
+			// close pending attribute blocks:
+			while( m_attrBlockCounter )
```

Comment:
Is it impossible to use a stack of `IECore::AttributeBlock` (which will do their own cleanup) instead of this? I'm guessing maybe it is because it's noncopyable?